File: src/server/aggregators/feed_content.ts
/**
 * RSS-Only aggregator.
 *
 * Lightweight aggregator that uses content directly from the RSS feed
 * without fetching full articles from the web.
 */

import Parser from "rss-parser";

import { BaseAggregator } from "./base/aggregator";
import { fetchFeed } from "./base/fetch";
import type { RawArticle } from "./base/types";

export class FeedContentAggregator extends BaseAggregator {
  override readonly id: string = "feed_content";
  override readonly type: "managed" | "custom" | "social" = "custom";
  override readonly name: string = "RSS-Only";
  override readonly url: string = "";
  override readonly description: string =
    "RSS feeds with full content already included in the feed.";
  override readonly identifierEditable: boolean = true;
  override readonly prefillName: boolean = false;

  /**
   * Fetch RSS feed data.
   */
  protected override async fetchSourceData(
    limit?: number,
  ): Promise<Parser.Output<unknown>> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
        limit,
      },
      "Fetching RSS feed",
    );

    if (!this.feed) {
      throw new Error("Feed not initialized");
    }

    const feedUrl = this.feed.identifier;
    const feed = await fetchFeed(feedUrl);

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "fetchSourceData",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        itemCount: feed.items?.length || 0,
        elapsed,
      },
      "RSS feed fetched",
    );

    return feed;
  }

  /**
   * Parse RSS feed items to RawArticle[].
   */
  protected override async parseToRawArticles(
    sourceData: unknown,
  ): Promise<RawArticle[]> {
    const startTime = Date.now();
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "start",
        aggregator: this.id,
        feedId: this.feed?.id,
      },
      "Parsing RSS feed items",
    );

    const feed = sourceData as Parser.Output<unknown>;
    const items = feed.items || [];

    const articles: RawArticle[] = items.map((item) => ({
      title: item.title || "",
      url: item.link || "",
      published: item.pubDate ? new Date(item.pubDate) : new Date(),
      content: item.content || item.contentSnippet || "",
      summary: item.contentSnippet || item.content || "",
      author: item.creator || (item as Parser.Item & { author?: string }).author || undefined,
    }));

    const elapsed = Date.now() - startTime;
    this.logger.info(
      {
        step: "parseToRawArticles",
        subStep: "complete",
        aggregator: this.id,
        feedId: this.feed?.id,
        articleCount: articles.length,
        elapsed,
      },
      "Parsed RSS feed items",
    );

    return articles;
  }

  /**
   * Check if content should be fetched - RSS-only aggregator never fetches.
   */
  protected override shouldFetchContent(_article: RawArticle): boolean {
    // RSS-only aggregator uses content from feed, never fetches from web
    return false;
  }

  /**
   * Process content from RSS feed (sanitize and standardize).
   */
  protected override async processContent(
    html: string,
    article: RawArticle,
  ): Promise<string> {
    // Use base processContent which handles sanitization and standardization
    return await super.processContent(html, article);
  }
}

File: legacy_backend/aggregators/feed_content.py
"""
Simple aggregator for RSS feeds.

This module provides a simple aggregator that uses the RSS feed content directly
without fetching the full article from the web. It sanitizes the content and
adds a source link.
"""

from pydantic import BaseModel, Field

from .base import BaseAggregator, RawArticle


class SimpleAggregatorConfig(BaseModel):
    id: str
    type: str = Field(pattern="^(managed|custom|social)$")
    name: str = Field(min_length=1)
    url: str = ""
    description: str = Field(min_length=1)
    wait_for_selector: str | None = None
    selectors_to_remove: list[str] = Field(default_factory=list)
    options: dict = Field(default_factory=dict)


class SimpleAggregator(BaseAggregator):
    """
    Lightweight aggregator that uses content directly from the RSS feed.

    Does not fetch full articles from the web - faster but may have incomplete content.
    Best for feeds with full content already included.
    """

    id = "feed_content"
    type = "custom"
    name = "RSS-Only"
    url = ""
    description = "Lightweight aggregator that uses content directly from the RSS feed without fetching full articles. Faster but may have incomplete content. Best for feeds with full content already included."

    def __init__(self):
        super().__init__()
        SimpleAggregatorConfig(
            id=self.id,
            type=self.type,
            name=self.name,
            url=self.url,
            description=self.description,
            wait_for_selector=self.wait_for_selector,
            selectors_to_remove=self.selectors_to_remove,
            options=self.options,
        )

    def fetch_article_html(self, article: RawArticle) -> str:
        """Don't fetch from web - use RSS content directly."""
        return article.content

    def extract_content(self, article: RawArticle) -> None:
        """RSS content is already extracted - nothing to do."""
        pass


def aggregate(feed, force_refresh=False, options=None):
    aggregator = SimpleAggregator()
    return aggregator.aggregate(feed, force_refresh, options or {})
